src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -1069,6 +1092,36 @@ private bool CheckPropertyValueKind(SyntaxNode node, BoundExpression expr, BindV
             }
 
             return true;
+
+            bool isAllowedInitOnlySet(BoundExpression receiver)
+            {
+                // ok: new C() { InitOnlyProperty = ... }
+                if (receiver is BoundObjectOrCollectionValuePlaceholder)
+                {
+                    return true;
+                }
+
+                // bad: other.InitOnlyProperty = ...
+                if (!(receiver is BoundThisReference || receiver is BoundBaseReference))
+                {
+                    return false;
+                }
+
+                var containingMember = ContainingMemberOrLambda;
+                if (!(containingMember is MethodSymbol method))
+                {
+                    return false;
+                }
+
+                bool isInitOnlySetter = method.MethodKind == MethodKind.PropertySet && method.IsInitOnly;
+                if (method.IsConstructor() || isInitOnlySetter)
```

Comment:
>IsConstructor [](start = 27, length = 13)

It looks like this check returns true for a static constructor. The spec doesn't allow assignment in a static constructor. Even though some other error is likely to be reported in this scenario, I think it would be better to use more precise condition that follows the spec "to the letter". Also, please add a test that gets here for a static constructor, even if that would be an error due to reasons unrelated to this feature.

Reply:
> It looks like this check returns true for a static constructor. The spec doesn't allow assignment in a static constructor. Even though some other error is likely to be reported in this scenario, I think it would be better to use more precise condition that follows the spec "to the letter". Also, please add a test that gets here for a static constructor, even if that would be an error due to reasons unrelated to this feature. 

This thread was marked as resolved. However, I don't see any changes for the corresponding code and no other response. If the context for the thread is not clear, it is about the ```method.IsConstructor()``` check.

---
In reply to: [415103737](https://github.com/dotnet/roslyn/pull/43275#discussion_r415103737) [](ancestors = 415103737)

Reply:
The resolution was the added because assertion above (`!method.IsStatic`).
Based on your comment above, sounds like you prefer an explicit check even if unreachable. I'm okay with that. Will change.

---
In reply to: [416728343](https://github.com/dotnet/roslyn/pull/43275#discussion_r416728343) [](ancestors = 416728343,415103737)

Reply:
>The resolution was the added because assertion above (!method.IsStatic).
Based on your comment above, sounds like you prefer an explicit check even if unreachable. I'm okay with that. Will change.

Given that the explicit check is simple and even less expensive than  ```method.IsConstructor()``` check, it feels reasonable to adjust the check rather than adding an assert, which would "save" us only we ran into a bad situation during testing.

---
In reply to: [416822362](https://github.com/dotnet/roslyn/pull/43275#discussion_r416822362) [](ancestors = 416822362,416728343,415103737)